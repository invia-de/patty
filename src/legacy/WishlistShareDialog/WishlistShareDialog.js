import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import copy from 'copy-to-clipboard';

import {
  Share,
  Close,
  WhatsApp,
  Facebook,
  Copy,
  Envelope,
  Tick
} from '../../components/atoms/Icon/Icon';
import Spinner from '../../components/atoms/Spinner/Spinner';
import Modal from '../../components/molecules/Modal/Modal';
import cx from '../../utils/classnames';
import track from '../../utils/ga-track';
import localStorageIsAvailable from '../../utils/localstorage';

import styles from './wishlistsharedialog.module.scss';

class WishlistShareDialog extends React.Component {
  constructor() {
    super();
    this.state = {
      isSaving: false,
      shareId: null,
      isCopied: false
    };
    this.modalRef = React.createRef();
    this.saveWishlist = this.saveWishlist.bind(this);
    this.actionCopy = this.actionCopy.bind(this);
    this.actionFacebook = this.actionFacebook.bind(this);
  }

  render() {
    return (
      <div>
        {this.renderTrigger()}
        {this.renderModal()}
      </div>
    );
  }

  renderTrigger() {
    const { children } = this.props;
    const { isSaving } = this.state;

    return (
      <button
        className={cx(styles.trigger, isSaving && styles.saving)}
        onClick={this.saveWishlist}
      >
        <span className={styles.triggerText}>
          {children} <Share />
        </span>
        <Spinner className={styles.spinner} />
      </button>
    );
  }

  renderModal() {
    const { portalName, eventAction } = this.props;
    const { isCopied } = this.state;

    const whatsappText = `Hallo, schau mal was ich tolles bei ${portalName} gefunden habe`;
    const whatsappLink = this.shareableLink('whatsapp');

    return (
      <Modal
        trigger={null}
        ref={ref => {
          this.modalRef = ref;
        }}
        isStatic
        className={styles.modal}
        onOpen={() =>
          track({
            event: 'gaEvent',
            eventCategory: 'Merkzettel',
            eventAction
          })
        }
      >
        <h1>
          <span>Merkzettel teilen</span>
          <button
            className={styles.mobileCloseButton}
            onClick={() => this.modalRef && this.modalRef.closeModal()}
            aria-label="Abbrechen"
          >
            <Close />
          </button>
        </h1>
        <div className={styles.options}>
          {this.renderOption(
            <WhatsApp />,
            'Mit Whatsapp teilen',
            () => this.trackShare('Mit Whatsapp teilen'),
            null,
            {
              href: `whatsapp://send?text=${encodeURIComponent(
                `${whatsappText}: ${whatsappLink}`
              )}`,
              'data-text': whatsappText,
              'data-href': whatsappLink
            },
            true
          )}
          {this.renderOption(
            <Facebook />,
            'Mit Facebook teilen',
            this.actionFacebook
          )}
          {this.renderOption(
            <Copy className={styles.aquaIcon} />,
            isCopied ? 'Link kopiert' : 'Als Link kopieren',
            this.actionCopy,
            isCopied ? <Tick /> : null
          )}
          {this.renderOption(
            <Envelope className={styles.aquaIcon} />,
            'Als E-Mail versenden',
            () => this.trackShare('Als E-Mail versenden'),
            null,
            {
              href: this.actionMail()
            }
          )}
        </div>
      </Modal>
    );
  }

  renderOption(
    icon,
    text,
    onClick = () => {},
    secondIcon = null,
    props,
    mobileOnly
  ) {
    return (
      <a
        onClick={onClick}
        alt={text}
        className={cx(styles.option, mobileOnly && styles.mobileOnly)}
        {...props}
      >
        {icon}
        <span>
          {text}
          {secondIcon}
        </span>
      </a>
    );
  }

  shareableLink(type) {
    const { baseURL, portalName } = this.props;
    const { shareId } = this.state;
    const trackParams = `${
      baseURL.includes('?') ? '&' : '?'
    }utm_source=${portalName}&utm_medium=referral&utm_campaign=wishlist_sharingdialog_${type}`;

    return baseURL + shareId + trackParams;
  }

  actionFacebook() {
    const link = new URL('https://www.facebook.com/sharer/sharer.php');
    link.search = new URLSearchParams({
      u: this.shareableLink('facebook')
    });
    this.trackShare('Mit Facebook teilen');
    window.open(link);
  }

  actionMail() {
    const { portalName } = this.props;
    const encoded_subject = encodeURIComponent(
      `Tolle Angebote auf ${portalName}`
    );
    const encoded_body = encodeURIComponent(
      `Hallo, schau mal was ich tolles bei ${portalName} gefunden habe: \n\n${this.shareableLink(
        'mail'
      )}`
    );

    return 'mailto:?subject=' + encoded_subject + '&body=' + encoded_body;
  }

  actionCopy() {
    copy(this.shareableLink('copy'));
    this.setState({ isCopied: true });
    this.trackShare('Link kopieren');
    setTimeout(() => this.setState({ isCopied: false }), 2000);
  }

  saveWishlist() {
    const { saveURL, agent, storageName } = this.props;
    const { isSaving } = this.state;

    if (isSaving || !localStorageIsAvailable) return;

    this.setState({ isSaving: true });

    const data = JSON.parse(window.localStorage.getItem(storageName)) || {};
    const wishlist = Object.keys(data).map(key => {
      const item = data[key];
      const tokens = key.split('_');
      return Object.assign(item, {
        hotelId: tokens[0],
        hotelIdType: tokens[1]
      });
    });

    fetch(saveURL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ wishlist, agent })
    })
      .then(response => response.json())
      .catch(error => {
        console.error(error);
        return null;
      })
      .then(json => {
        const isSaving = false;
        const shareId = json.success ? json.response.id : null;
        this.setState({ isSaving, shareId }, () => {
          this.modalRef && this.modalRef.openModal();
        });
      });
  }

  trackShare(eventAction) {
    track({
      event: 'gaEvent',
      eventCategory: 'Merkzettel',
      eventAction
    });
  }
}
WishlistShareDialog.propTypes = {
  /** name of the storage to use */
  storageName: PropTypes.string,
  /** additional classNames you want to add */
  className: PropTypes.string,
  /** required due to accessibility */
  children: PropTypes.node,
  /** URL to be called to save a wishlist */
  saveURL: PropTypes.string,
  /** Agent to use when saving the shared wishlist */
  agent: PropTypes.string,
  /** The name of the curretn portal */
  portalName: PropTypes.string,
  /** Base wishlist URL where to append the shareID to when sharing */
  baseURL: PropTypes.string,
  /** Event name to fire when the sharing dialog is opened */
  eventAction: PropTypes.string
};

WishlistShareDialog.defaultProps = {
  storageName: 'wishlist',
  children: 'Teilen',
  baseURL: 'https://ab-in-den-urlaub.de/merkzettel?id=',
  agent: 'ab-in-den-urlaub.de',
  portalName: 'ab-in-de-urlaub.de',
  eventAction: 'Merkzettel teilen'
};

export default WishlistShareDialog;

export function renderWishlistShareDialog(props, container, callback) {
  ReactDOM.render(<WishlistShareDialog {...props} />, container, callback);
}
