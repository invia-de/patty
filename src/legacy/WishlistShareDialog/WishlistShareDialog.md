```jsx
<WishlistShareDialog
  agent="ab-in-den-urlaub.de"
  saveURL="https://mobileapi.test.invia.lan/FFP495/ms/v/5/service-wishlist-save"
/>
```
