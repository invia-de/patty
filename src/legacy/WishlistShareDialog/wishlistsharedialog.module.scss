@import 'theme.scss';

$wishlistsharedialog-bgcolor: $aqua !default;
$wishlistsharedialog-color: $white !default;
$wishlistsharedialog-modal-color: $blue !default;
$wishlistsharedialog-modal-border: #adc6f2 !default;

button.trigger,
button.trigger:focus,
button.trigger:hover {
  position: relative;
  height: 24px;
  font-size: 14px;
  line-height: 16px;
  padding: 4px 11px;
  background-color: $wishlistsharedialog-bgcolor !important;
  border-radius: 2px;
  cursor: pointer;
  &,
  & > span {
    color: $wishlistsharedialog-color !important;
  }
  svg {
    position: relative;
    top: -2px;
  }
  .spinner {
    opacity: 0;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 42px;
    & > div {
      width: 12px;
      height: 12px;
      background-color: $wishlistsharedialog-color;
    }
  }
  &.saving {
    .triggerText {
      opacity: 0;
    }
    .spinner {
      opacity: 1;
    }
  }
}

.modal {
  min-width: 290px;
  overflow-x: hidden;
  padding: 15px !important;
  h1 {
    display: flex !important;
    justify-content: space-between !important;
    color: $wishlistsharedialog-modal-color !important;
    font-size: 22px !important;
    font-weight: bold;
    margin-bottom: 10px !important;
    button,
    button:focus,
    button:hover {
      font-size: 14px;
      position: relative;
      top: -15px;
      left: 16px;
      height: auto;
      line-height: auto;
      padding: 10px;
      color: $wishlistsharedialog-modal-color !important;
      background-color: transparent;
      border-radius: 0;
    }
    span {
      color: $wishlistsharedialog-modal-color !important;
    }
  }
}

.options {
  color: $wishlistsharedialog-modal-color;
  user-select: none;
}

.option {
  display: flex;
  width: 100%;
  padding: 15px 0;
  font-size: 16px;
  color: #2a4d8f !important;
  text-align: left;
  &,
  & > * {
    cursor: pointer;
  }
  svg {
    margin-right: 15px;
  }
  span {
    flex-grow: 1;
    color: $wishlistsharedialog-modal-color !important;
    svg {
      position: relative;
      left: 5px;
      top: -3px;
    }
  }
}

.option:not(:last-child) {
  border-bottom: 1px solid $wishlistsharedialog-modal-border;
}

svg.aquaIcon path {
  fill: $wishlistsharedialog-bgcolor;
}

.mobileOnly {
  display: none;
}

@include respond-below(md) {
  .mobileOnly {
    display: flex;
  }
}
