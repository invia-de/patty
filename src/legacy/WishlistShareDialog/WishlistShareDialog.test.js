import React from 'react';
import WishlistShareDialog from './WishlistShareDialog';
import { render } from '@testing-library/react';

test('WishlistShareDialog renders', () => {
  const { container } = render(<WishlistShareDialog />);
  expect(container.firstChild).toMatchSnapshot();
});
