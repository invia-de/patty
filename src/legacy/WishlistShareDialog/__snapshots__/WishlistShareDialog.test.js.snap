// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishlistShareDialog renders 1`] = `
<div>
  <button
    class="trigger"
  >
    <span
      class="triggerText"
    >
      Teilen
       
      <svg
        aria-hidden="true"
        class="icon share"
        viewBox="0 0 13 14"
      >
        <path
          d="M10.801 9.896c-.547 0-1.036.21-1.409.54L4.256 7.52c.038-.161.066-.324.066-.492 0-.168-.03-.33-.066-.492l5.078-2.888c.387.35.9.568 1.467.568 1.197 0 2.162-.94 2.162-2.108C12.963.942 11.998 0 10.801 0c-1.194 0-2.16.942-2.16 2.108 0 .17.03.33.066.492L3.63 5.49c-.388-.351-.9-.57-1.468-.57C.965 4.92 0 5.862 0 7.028c0 1.166.965 2.108 2.162 2.108a2.18 2.18 0 0 0 1.467-.569l5.127 2.922a2 2 0 0 0-.058.459c0 1.131.945 2.052 2.103 2.052 1.159 0 2.104-.92 2.104-2.052 0-1.132-.945-2.052-2.104-2.052z"
        />
      </svg>
    </span>
    <div
      class="spinner spinner"
    >
      <div
        class="bounce1"
      />
      <div
        class="bounce2"
      />
      <div
        class="bounce3"
      />
    </div>
  </button>
  <div>
    <div
      class="trigger"
    />
  </div>
</div>
`;
